nft-marketplace/src/sale.rs

```diff
@@ -1,27 +1,23 @@
-use crate::{nft_messages::*, payment::*, Market, MarketEvent, BASE_PERCENT};
-use gstd::{msg, prelude::*, ActorId};
-use primitive_types::{H256, U256};
+use crate::{
+    assert_auction_is_on, get_item, nft_messages::*, payment::*, ContractId, Market, MarketEvent,
+    TokenId, BASE_PERCENT,
+};
+use gstd::{msg, prelude::*};
 
 impl Market {
-    pub async fn buy_item(&mut self, nft_contract_id: &ActorId, token_id: U256) {
-        let contract_and_token_id =
-            format!("{}{}", H256::from_slice(nft_contract_id.as_ref()), token_id);
-        let item = self
-            .items
-            .get_mut(&contract_and_token_id)
-            .expect("Item does not exist");
-        if item.auction.is_some() {
-            panic!("There is an opened auction");
-        }
+    pub async fn buy_item(&mut self, nft_contract_id: ContractId, token_id: TokenId) {
+        let item = get_item(&mut self.items, nft_contract_id, token_id);
+        assert_auction_is_on(&item.auction);
         let price = item.price.expect("The item is not on sale");
 
         check_attached_value(item.ft_contract_id, price);
         // fee for treasury
-        let treasury_fee = price * (self.treasury_fee * BASE_PERCENT) as u128 / 10_000u128;
+        let treasury_fee =
+            price * (self.treasury_fee as u16 * BASE_PERCENT as u16) as u128 / 10_000u128;
```

Comment:
`treasury_fee` and `BASE_PERCENT` are already `u16`'s and you should check multiplication for overflowing I think or make them u32